ddl/partition.go

```diff
@@ -1115,20 +1115,31 @@ func setPartitionPlacementFromOptions(partition *model.PartitionDefinition, opti
 	return nil
 }
 
-func buildHashPartitionDefinitions(_ sessionctx.Context, defs []*ast.PartitionDefinition, tbInfo *model.TableInfo) ([]model.PartitionDefinition, error) {
+func buildHashPartitionDefinitions(_ sessionctx.Context, defs []*ast.PartitionDefinition, tbInfo *model.TableInfo, numParts uint64) ([]model.PartitionDefinition, error) {
 	if err := checkAddPartitionTooManyPartitions(tbInfo.Partition.Num); err != nil {
 		return nil, err
 	}
 
-	definitions := make([]model.PartitionDefinition, tbInfo.Partition.Num)
-	for i := 0; i < len(definitions); i++ {
-		if len(defs) == 0 {
-			definitions[i].Name = model.NewCIStr(fmt.Sprintf("p%v", i))
-		} else {
+	definitions := make([]model.PartitionDefinition, numParts)
+	if len(defs) == 0 {
+		for i := 0; i < len(definitions); i++ {
+			definitions[i].Name = model.NewCIStr(fmt.Sprintf("p%d", i))
+		}
+		return definitions, nil
```

Comment:
I'm afraid this will have some problem:

```sql
create table t (id int) partition by hash (id) partitions 5;
alter table t add partition (partition pt comment 'pt');
alter table t add partition partitions 2;
show create table t\G   -- The output is not expected

> ***************************[ 1. row ]***************************
Table        | t
Create Table | CREATE TABLE `t` (
  `id` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY HASH (`id`) PARTITIONS 8
```

Reply:
Ah, I must have missed the ADD case and only checked the COALESCE case, looking into it.

Reply:
Great catch! I have updated the PR.